[WIP] Multi-arch build-push helper script

cevich opened the pull request:
This will partially address https://github.com/containers/buildah/issues/3268 but the initial intent is to include this script in podman/buildah/skopeo automation.  So there may be some initial end-user unfriendliness.

edsantiago reviewed (changes requested):
First pass, but not a 100% thorough one (I got tired!)

cevich commented:
Thanks @edsantiago for so many helpful suggestions, especially on the CLI option expectations and related docs.  

> not a 100% thorough one (I got tired!)

 Nono, this is great.  I just needed an initial sanity-check over exactly what you looked at.  I was afraid of getting too far and needing to re-architect due to some dumb choice (like if `--tagcmd` was a totally dumb idea).  Thanks a bunch, I'll make all those fixes and press onward.

cevich commented:
Ahh, lovely, see there, flake happened and proves 20-minute time is helpful :grin: 
https://cirrus-ci.com/task/5621331594051584

cevich commented on build-push/bin/build-push.sh:
```diff
@@ -0,0 +1,373 @@
+                ;;
+            --nopush)
+                dbg "Nopush flag detected, will NOT push built images."
+                NOPUSH=0  # i.e. "Yes, do not push"
```
> Nooooo.... please no

Oof...having to guess what you mean by that :smile:   I already re-implemented this to use only positive logic, storing `1` in `$PUSH` when the `--nopush` flag is given.  That's ***MUCH*** easier to read/understand.  However, even this could possibly be simplified further.

I think (tell me if I'm wrong), when CI is running on a branch and intended to push images, it will explicitly be given access to the secrets.   Otherwise I think it's a safe assumption that "no-password" == the caller just wants to build locally.  

Would that be a worthwhile further improvement?

cevich replied:
Clearification:  "no-password" meaning `$<NAMESPACE>_PASSWORD` is empty or unset.

edsantiago replied:
What I meant was, if you're going to use `NOPUSH` as a variable name, it should be set to 1 when `--nopush` is given. `NOPUSH=0` in the above context is Chaotic Evil.

I don't have a good enough sense for CI secrets management to reply to your question. Is it possible for secrets to silently go away? In that case the job would succeed without pushing, probably for a very long time without anyone noticing. If that's possible/likely, I think it might make sense to keep it as-is (fail loudly if secrets are missing). If it's unlikely, then I'm OK with your judgment call either way.

cevich commented on build-push/README.md:
```diff
@@ -0,0 +1,91 @@
+
+* Automation common-library (installed with bin/install_automation.sh)
+* Executables for `jq`, and `buildah` are available.
+* (Optional) The kernel is configured to use QEMU emulation for
```
Generally yes, but I didn't want to force / lock-in on this particular emulation technique.  It's not available in RHEL or CentOS  nor is it applicable on all architectures.  e.g. x86_64 can execute x86 code w/o any emulation (assuming the right bits/libraries are available).  There are similar native execution modes available to s390/s390x, ppc64/ppc64le, and likely others.

cevich commented on build-push/README.md:
````diff
@@ -0,0 +1,91 @@
+    --tagcmd=./path/to/tag_script.sh
+```
+
+Here, `tag_script.sh` contains the logic necessary for extracting/producing
+a useful tag.  It will only be executed ***after*** a successful build of
````
This is better explained in the `--help` output, you're right it should be in the example here.

cevich commented on build-push/bin/build-push.sh:
```diff
@@ -0,0 +1,373 @@
+                 sed -z '$ s/[\n ]$//')
+    dbg "Found manifest arches: $maniarches"
+
+    for maniarch in $maniarches; do
```
good catch.

cevich commented:
Major changes:

* `s/bud/build/g` everywhere
* Killed off `--tagcmd` completely, since it can easily be done as part of `--modcmd` as needed.
* Renamed `$NOPUSH` to `$PUSH` along with other related readability improvements
* Lots of `README.md` fixes and clarified examples
* Lots of `--help` output cleanup
* Re-worked registry login function to be called from multiple places w/o logging in multiple times
* Implemented pushing function
* Reworked tests as needed to pass (one test is temporarily disabled due to buildah open issue)

cevich commented:
@edsantiago please don't make another pass yet, I need to work it over a few more times myself first.  I'm positive there are a few "it's" I missed :grinning:

cevich commented:
More changes:
* Added "e2e" test with an actual multi-arch build and push to a test registry.
* Additional `README.md` cleanup/fixes
* Added a "stage notice" function to help humans distinguish pre, build, and post output.  Otherwise it's a massive wall-of-text.
* Fixed a typo of `NAMESPASCE_USERNAME`
* Rearchitected use of `$MANIFESTID` - not reliable to use in practice, need to use the FQIN for manifest-lists.
* Implemented workaround (for testing only!) to install qemu-user-static in CentOS stream-8 VM

cevich commented:
Pre-review final cleanup:
* Added check on e2e test build, confirming all expected arches are present.
* Moved/clarified comments in `.cirrus.yml`.
* Switched to using a buildah namespace instead of a personal one for the test-push.
* Pass in test FQIN from `.cirrus.yml` into test, placing definition close to needed credentials.
* Added checks for required variables in `testbuilds.sh`.
* `build-push/README.md` fixes/adjustments.

cevich commented:
@edsantiago PTAL.  Hopefully my changes summaries are helpful.  The test-suite isn't perfect, but should be good n'uff for starting out.  I plan to add more tests when incorporating this into a VM image for actual use in containers-org projects.

edsantiago reviewed (commented):
I only reviewed `range-diff`, but even so that was a lot: I ran out of steam near the end (tests).

LGTM, my comments are mostly typos. @rhatdan has a tool somewhere that does spellchecks in code, maybe you could add that to your pre-commit habits?

cevich commented:
> I ran out of steam near the end (tests).

Still, I appreciate all the effort.  The tests setup here is less than ideal and quite hacky given they need to do a lot with very little in the way of pre-installed tooling (i.e. none).  There'll be more practical testing done at higher-levels when I work this script up the automation-food-chain. 

> @rhatdan has a tool somewhere that does spellchecks in code, maybe you could add that to your pre-commit habits?

Oh!  That's a novel idea, thanks!

cevich commented on build-push/bin/build-push.sh:
```diff
@@ -0,0 +1,445 @@
+
+if [[ ! -r "$AUTOMATION_LIB_PATH/common_lib.sh" ]]; then
+    (
+        echo "ERROR: Expecting \$AUTOMATION_LIB_PATH to be set.  Normally"
```
Fixed/clarified and added reference to RTFM `README.md`.
